ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,545 @@
+    bsClose = do
+      Trace.traceWith tracer $ TDBClosing dbFilePath
+      liftIO $ do
+        open_handles <- IOLike.atomically $ IOLike.readTVar dbOpenHandles
+        for_ open_handles vhClose
+        IOLike.atomically $ IOLike.writeTBQueue trace_q Nothing
+      IOLike.wait trace_async
+      liftIO $ LMDB.closeEnvironment dbEnv
+      Trace.traceWith tracer $ TDBClosed dbFilePath
+
+    bsCopy shfs (HD.BackingStorePath to0) = do
+      to <- guardDbDir GDDMustNotExist shfs to0
+      liftIO $ lmdbCopy dbTracer dbEnv to
+
+    bsValueHandle = mkLMDBBackingStoreValueHandle db
+
+    bsWrite :: SlotNo -> LedgerTables l DiffMK -> m ()
+    bsWrite slot diffs = liftIO $ do
+      old_slot <- LMDB.readWriteTransaction dbEnv $ withDbSettingsRW dbSettings $ \s@DbState{dbsSeq} -> do
+        -- TODO This should be <. However the test harness does call bsWrite with the same slot
+        unless (dbsSeq <= At slot) $ Exn.throw $ DbErrNonMonotonicSeq (At slot) dbsSeq
+        void $ zipLedgerTablesA writeLMDBTable dbBackingTables diffs
+        pure (dbsSeq, s {dbsSeq = At slot})
+      Trace.traceWith dbTracer $ TDBWrite old_slot slot
+
+  -- now initialise those tables if appropriate
+  init_action db
+
+  Trace.traceWith tracer $ TDBOpened dbFilePath
+  pure HD.BackingStore{..}
+
+
+data SomeDbSubmission where
+  SomeDbSubmission :: forall a. LMDB.Transaction LMDB.ReadOnly a -> IOLike.TMVar IO (Maybe a) -> SomeDbSubmission
+
+data TraceValueHandle
+  = TVHOpened
+  | TVHClosing
+  | TVHClosed
+  | TVHSubmissionStarted
+  | TVHSubmissionEnded
+  deriving stock(Show, Eq)
+
+mkValueHandle :: (MonadIO m)
+  => Trace.Tracer IO TraceDb
+  -> LMDB.Environment LMDB.ReadOnly
+  -> IOLike.TVar IO (Map Int ValueHandle)
+  -> m ValueHandle
+mkValueHandle tracer0 dbEnv dbOpenHandles = liftIO $ do
+  q <- IOLike.newTBQueueIO 1 -- TODO 1?
+  the_async <- IOLike.async $ LMDB.readOnlyTransaction dbEnv $ do
+    let
+      loop :: MaybeT (LMDB.Transaction LMDB.ReadOnly) ()
+      loop = do
+        -- TODO It seems much time is spent in this (and other) atomically calls
+        SomeDbSubmission t ret_tmvar <- MaybeT . liftIO $ IOLike.atomically (IOLike.readTBQueue q)
+        r <- lift t
+        liftIO . IOLike.atomically $ IOLike.putTMVar ret_tmvar (Just r)
+    void . runMaybeT $ forever loop
+    LMDB.abort
+
+  (r, traces) <- IOLike.atomically $ IOLike.stateTVar dbOpenHandles $ \x0 -> let
+
+    vh_id = maybe 0 ((+1) . fst) $ Map.lookupMax x0
+    tracer = Trace.contramap (TDBValueHandle vh_id) tracer0
+
+    vhSubmit :: forall a. LMDB.Transaction LMDB.ReadOnly a -> IO (Maybe a)
+    vhSubmit t = do
+      Trace.traceWith tracer TVHSubmissionStarted
+      ret_tmvar <- IOLike.newEmptyTMVarIO
+      IOLike.atomically . IOLike.writeTBQueue q . Just $ SomeDbSubmission t ret_tmvar
+      r <- IOLike.atomically $ IOLike.takeTMVar ret_tmvar
+      Trace.traceWith tracer TVHSubmissionEnded
+      pure r
+
+    vhClose = do
+      Trace.traceWith tracer TVHClosing
```

Comment:
I have a feeling that we are doing a bit of an overkill with so much tracing in this module.